### PR TITLE
Refactor config API

### LINKED ISSUE #4989: [StandardLib] Config API refactor

### LINKED ISSUE #5206: [STDLIB] Unable to override config values
**Description:**
Unable to override config values with environment variables. Getting the error:

ballerina: internal error occurred

**Steps to reproduce:**
```
hemikakodikara@Hemikas-MacBook-Pro:central-test$ export ABCD=YYYYYY
hemikakodikara@Hemikas-MacBook-Pro:central-test$ ballerina run test.bal -Bballerina.conf=config.conf -Babcd=ZZZZZ
ballerina: internal error occurred
```

**Affected Versions:**
0.964.0

**OS, DB, other environment details and versions:**
MAC

### PR DESCRIPTION
**Pull request by pubudu91**
## Samples
```ballerina
import ballerina.config;
import ballerina.io;

function main(string[] args) {
    string[] users = config:getAsString("username.instances").split(",");
    string user1Rights = config:getAsString(users[0] + ".access.rights");
    string user2Rights = config:getAsString(users[1] + ".access.rights");

    io:println(users[0] + " has " + user1Rights + " access");
    io:println(users[1] + " has " + user2Rights + " access");

    io:println("Before changing sum.limit in code: " + sum());
    config:setConfig("sum.limit", "10");
    io:println("After changing sum.limit: " + sum());
}

function sum () (int) {
    var to, err = <int>config:getAsString("sum.limit");

    if (err != null) {
        return -1; // Returning -1 to signal failure
    }

    int sumUpTo = 0;
    int i = 1;

    while (i <= to) {
        sumUpTo = sumUpTo + i;
        i = i + 1;
    }

    return sumUpTo;
}

```
The config file:
```toml
sum.limit=5
username.instances="john,peter"
[john]
access.rights="RW"
[peter]         	
access.rights="R"
```
To run the program with the given config file:
```
$ ballerina run --config <PATH_TO_FILE>/ballerina.conf config-api-example.bal
```

Configuration values can be encrypted values as well. 
```
user.password="@encrypted:{UtD9d+o6eHpqFnBxtvhb+RWXey7qm7xLMt6+6mrt9w0=}"
```
Sensitive values can be encrypted using the `encrypt` command as follows. The user will be prompted to enter the value to be encrypted and a secret to be used during encryption.
```$ ballerina encrypt
ballerina: enter the value to be encrypted:

ballerina: enter a secret to be used in encrypting the value:

A6mSL9rZbG9WqT32eW3PUA/L9nxylcp5QgIRPmNXUsE=
```

Decryption of encrypted values is automatically handled by the config API. However the user needs to provide the secret used to encrypt the values with. There are 2 options for the user for this.
- Provide a `secret.txt` file which contains the secret. This needs to be placed in the source root and should be named as shown. If placing the secret file at a different location, that needs to be passed as a configuration (either through a file or as a CLI param): `ballerina.config.secret`
- If the user hasn't provided a secret file and the provided configurations contain encrypted values, Ballerina will prompt the user to type the secret.

Users can specify environment variable keys as `@env:{<ENV_VAR_NAME>}`. When reading the config file/CLI params, the config registry will resolve and store the value of `<ENV_VAR_NAME>`. 
e.g: `user.path="@env:{PATH}"`

For Ballerina developers, an API is provided for reading/setting VM options. Developers can access this API from the `VMOptions` singleton.

### CODE REVIEW

File: stdlib/ballerina-config/src/main/java/org/ballerinalang/bcl/parser/antlr4/TomlLexer.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	}
+
+	@Override
+	@Deprecated
+	public String[] getTokenNames() {
+		return tokenNames;
+	}
+
+	@Override
+
```
Comment (VijithaEkanayake) [inline stdlib/ballerina-config/src/main/java/org/ballerinalang/bcl/parser/antlr4/TomlLexer.java]: Formatting issue??

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java [hunk 2]
```diff
@@ -415,6 +424,52 @@ public void setSelfCmdParser(JCommander selfCmdParser) {
...
+            String value = new String(System.console().readPassword());
+
+            outStream.println("ballerina: enter a secret to be used in encrypting the value:");
+            String secret = new String(System.console().readPassword());
+
+            try {
+                AESCipherTool cipherTool = new AESCipherTool(secret);
+                outStream.println(cipherTool.encrypt(value));
+            } catch (AESCipherToolException e) {
+                e.printStackTrace();
```
Comment (chamil321) [inline cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: Shall we log this properly?

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java [hunk 3]
```diff
@@ -415,6 +424,74 @@ public void setSelfCmdParser(JCommander selfCmdParser) {
...
+        @Override
+        public void execute() {
+            if (helpFlag) {
+                String commandUsageInfo = BLauncherCmd.getCommandUsageInfo(parentCmdParser, "encrypt");
+                outStream.println(commandUsageInfo);
+                return;
+            }
+
+            outStream.println("ballerina: enter the value to be encrypted:");
+            String value = new String(System.console().readPassword());
```
Comment (ayomawdb) [inline cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: Should we allow encrypting an empty string? We should ideally re-prompt if user did not enter anything.

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java [hunk 4]
```diff
@@ -415,6 +424,74 @@ public void setSelfCmdParser(JCommander selfCmdParser) {
...
+
+            outStream.println("ballerina: enter a secret to be used in encrypting the value:");
+            String secret = new String(System.console().readPassword());
```
Comment (ayomawdb) [inline cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: Should we allow using an empty password? We should ideally re-prompt if user did not enter anything.

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java [hunk 5]
```diff
@@ -415,6 +424,74 @@ public void setSelfCmdParser(JCommander selfCmdParser) {
...
+        public String getName() {
+            return "encrypt";
+        }
+
+        @Override
+        public void printLongDesc(StringBuilder out) {
+            out.append("The encrypt command can be used to encrypt sensitive data.\n\n");
+            out.append("When the command is executed, the user will be prompted to enter the value to be encrypted ");
+            out.append("and a secret. The secret will be used in encrypting the value.\n\n");
+            out.append("Once encrypted, the user can place it in config files.\n");
```
Comment (ayomawdb) [inline cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: Shall we make it read : 
"Once encrypted, the user can place the encrypted value in the config files, similar to the following example:"

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java [hunk 6]
```diff
@@ -415,6 +424,74 @@ public void setSelfCmdParser(JCommander selfCmdParser) {
...
+            out.append("\te.g: user.password=\"@encrypted:{UtD9d+o6eHpqFnBxtvhb+RWXey7qm7xLMt6+6mrt9w0=}\"\n");
+        }
```
Comment (ayomawdb) [inline cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: Shall we also add: 
"Ballerina Config API will take care of automatically decryption the encrypted values on demand."

File: stdlib/ballerina-config/src/main/java/org/ballerinalang/bcl/parser/BConfigLangListener.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * A TOML parser listener implementation for parsing Ballerina config files.
+ *
+ * @since 0.965.0
+ */
+public class BConfigLangListener extends TomlBaseListener {
+
+    private static final String ENCRYPTED_FIELD_REGEX =
```
Comment (ayomawdb) [inline stdlib/ballerina-config/src/main/java/org/ballerinalang/bcl/parser/BConfigLangListener.java]: Did we write this on our own? If we did borrow the regex (part after "encrypted"), let's mention the reference in a comment.

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java (hunk 3 above)
Comment (pubudu91) [reply cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: Should we keep re-prompting? Or should we just re-prompt once and then exit if an empty string is added again?

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java (hunk 4 above)
Comment (pubudu91) [reply cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: Should we keep re-prompting? Or should we just re-prompt once and then exit if an empty string is added again?

File: stdlib/ballerina-config/src/main/java/org/ballerinalang/bcl/parser/BConfigLangListener.java (hunk 7 above)
Comment (pubudu91) [reply stdlib/ballerina-config/src/main/java/org/ballerinalang/bcl/parser/BConfigLangListener.java]: Borrowed. Will add a comment

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java [hunk 8]
```diff
@@ -415,6 +424,74 @@ public void setSelfCmdParser(JCommander selfCmdParser) {
...
+            if (helpFlag) {
+                String commandUsageInfo = BLauncherCmd.getCommandUsageInfo(parentCmdParser, "encrypt");
+                outStream.println(commandUsageInfo);
+                return;
+            }
+
+            outStream.println("ballerina: enter the value to be encrypted:");
+            String value = new String(System.console().readPassword());
+
+            outStream.println("ballerina: enter a secret to be used in encrypting the value:");
```
Comment (ayomawdb) [inline cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: This is how OpenSSL works: 
> enter aes-256-cbc encryption password:
> Verifying - enter aes-256-cbc encryption password:

Basically, it asks for password verification, before encrypting the value with (possibly misspelled) password. This is same with keytool passwords, etc. IMO, it is better to to the same.

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java (hunk 3 above)
Comment (ayomawdb) [reply cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: +1 to re-prompt once and then exit

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java (hunk 4 above)
Comment (ayomawdb) [reply cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: +1 to re-prompt once and then exit

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java (hunk 5 above)
Comment (pubudu91) [reply cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: ack.

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java (hunk 8 above)
Comment (pubudu91) [reply cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: ack.

File: cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java (hunk 6 above)
Comment (pubudu91) [reply cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java]: ack.

**Pull request merged**